src/main/scala/mesosphere/marathon/core/election/ElectionService.scala

```diff
@@ -91,10 +105,155 @@ trait ElectionCandidate {
   def startLeadership(): Unit
 }
 
-/** Local leadership events. They are not delivered via the event endpoints. */
-sealed trait LocalLeadershipEvent
+class ElectionServiceImpl(
+    eventStream: EventStream,
+    hostPort: String,
+    leaderEventsSource: Source[LeadershipState, Cancellable],
+    crashStrategy: CrashStrategy,
+    electionEC: ExecutionContext
+)(implicit system: ActorSystem) extends ElectionService with StrictLogging {
+
+  import ElectionService._
+  @volatile private[this] var lastState: LeadershipState = LeadershipState.Standby(None)
+  implicit private lazy val materializer = ActorMaterializer()
+  var leaderSubscription: Option[Cancellable] = None
+
+  def subscribe(subscriber: ActorRef): Unit = {
+    eventStream.subscribe(subscriber, classOf[LeadershipTransition])
+    val currentState = if (isLeader) LeadershipTransition.ElectedAsLeader else LeadershipTransition.Standby
+    subscriber ! currentState
+  }
+
+  def unsubscribe(subscriber: ActorRef): Unit = {
+    eventStream.unsubscribe(subscriber, classOf[LeadershipTransition])
+  }
+
+  override def isLeader: Boolean =
+    lastState == LeadershipState.ElectedAsLeader
 
-object LocalLeadershipEvent {
-  case object ElectedAsLeader extends LocalLeadershipEvent
-  case object Standby extends LocalLeadershipEvent
+  override def localHostPort: String = hostPort
+
+  override def leaderHostPort: Option[String] = lastState match {
+    case LeadershipState.ElectedAsLeader =>
+      Some(hostPort)
+    case LeadershipState.Standby(currentLeader) =>
+      currentLeader
+  }
+
+  /**
+    * Releases leadership
+    *
+    * Has no effect if called before offerLeadership
+    */
+  override def abdicateLeadership(): Unit = {
+    leaderSubscription.foreach(_.cancel())
+  }
+
+  private def initializeStream(leadershipTransitionsFlow: Flow[LeadershipState, LeadershipTransition, NotUsed]) = {
+    val localEventListenerSink = Sink.foreach[LeadershipState] { event =>
+      lastState = event
+    }
+    val localTransitionSink = Sink.foreach[LeadershipTransition] { e =>
+      eventStream.publish(e)
+      if (e == LeadershipTransition.Standby) {
+        logger.error("Lost leadership; crashing")
+        crashStrategy.crash()
+      }
+    }
+
+    val (leaderStream, leaderStreamDone) =
+      RunnableGraph.fromGraph(GraphDSL.create(
```

Comment:
How about moving this graph into another builder method so we can test it?

Reply:
Graph is already covered by tests. I don't feel like more is necessary.

Reply:
`leaderStream` is unit tested?